feat(contentUserData): add feature to save and retrieve contentUserdata (closes #1014)

JPSchellenberg opened the pull request:
This PR will introduce the feature of saving and retreiving contentUserData.

JPSchellenberg committed:
test(jest): run jest tests in watch mode via test:watch

JPSchellenberg committed:
chore(scripts): add start:rest script

JPSchellenberg committed:
feat(contentUserData): add parameters to contentUserDataUrl config

JPSchellenberg committed:
feat(contentUserData): read saveFreq from config

JPSchellenberg committed:
feat(contentUserData): add the contentUserData into the H5PIntegration

sr258 commented on packages/h5p-server/src/types.ts:
```diff
@@ -543,6 +544,48 @@ export interface IUser {
+    ): Promise<void>;
+
+    /**
+     * Loads the contentUserData for given contentId and builds an array of IContentUserData which is used in the integtration object of the rendered H5P by the H5PPlayer.
```
This method should not be part of the Storage interface but in the Manager class, as it applies to all storage classes.

JPSchellenberg replied:
The problem is that if it is only in the Manager class you have no idea/indication how many subContentIds there are.

Right now the implementation of `generateContentUserDataIntegration` the `IContentUserDataStorage` has to load all ContentUserData and build an array where the position in the array is equal to the subContentId.

If you use `loadContentUserData` from the Manager you don't know how many subContentIds / dataTypes there are. 

Or am I missing something?

sr258 replied:
In this case the interface should include a method like listByContent(contentId, userId) and the actual integration construction should be done in the Manager.

sr258 commented on packages/h5p-server/src/types.ts:
```diff
@@ -1533,6 +1576,13 @@ export interface IH5PConfig {
+     * the user object is passed into the render method of the player. You also
+     * must set ajax.contentUserData for state saving to work.
+     */
+    saveFreq: number | boolean;
```
I think the config option should be named in a a clearer way. The name Joubel chose is unclear and doesn't really help.

JPSchellenberg replied:
I agree. Any suggestions? I would suggest `contentUserDataSaveInterval` or `contentUserStateSaveInterval`.

sr258 replied:
I'd opt for contentUserStateSaveInterval.

sr258 commented on packages/h5p-server/src/H5PEditor.ts:
```diff
@@ -97,7 +100,8 @@ export default class H5PEditor {
         private urlGenerator: IUrlGenerator = new UrlGenerator(config),
-        private options?: IH5PEditorOptions
+        private options?: IH5PEditorOptions,
+        public contentUserDataStorage?: IContentUserDataStorage
```
The options object should always be the last parameter. I had to add the ILockProvider parameter in the other PR and I put it into the objects. I think it would be best to refactor the constructor to have parameters like this storage: {libraryStorage: ILibraryStorage, contentStorage: IContentStorage, ...}. That way we avoid too many breaking changes when adding new optional parameters.

JPSchellenberg replied:
Yeah. I put it last to avoid a breaking change since we basically only introduce a feature. I thought about the storage object as well and think that would be the best way to go.

sr258 replied:
The storage object is still a good idea.

JPSchellenberg replied:
But I would do that in another PR as it is not really related to this feature. We shouldn't do too many things in one PR.

sr258 commented on packages/h5p-server/src/H5PEditor.ts:
```diff
@@ -126,6 +130,9 @@ export default class H5PEditor {
             this.config
         );
+        this.contentUserDataManager = new ContentUserDataManager(
+            contentUserDataStorage
```
What happens when contentUserDataStorage is undefined?

JPSchellenberg replied:
The contentUserDataManger is constructed, but if the methods are called they return `undefined`:

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886/files#diff-175df6189e4a5c5418312fab8fcb2634717e3565726190172a885f90a9fd26e7R73

Should I use another way of doing this?

sr258 replied:
I see. I'd have implemented this in H5PEditor, but it can also be done in ContentUserDataManager

sr258 commented on packages/h5p-server/test/H5PPlayer.renderHtmlPage.test.ts:
```diff
@@ -39,7 +39,7 @@ describe('Rendering the HTML page', () => {
                         window.H5PIntegration = {
                   "ajax":{
-                     "contentUserData":"/h5p/contentUserData",
+                     "contentUserData":"/h5p/contentUserData/:contentId/:dataType/:subContentId",
```
We should consider moving to Jest Snapshots. Mainting this test file is quite annoying. I have touched it myself several times and it's very redundant work.

JPSchellenberg replied:
Ok, I made an issue for that: https://github.com/Lumieducation/H5P-Nodejs-library/issues/1892

sr258 commented on package.json:
```diff
@@ -36,6 +36,7 @@
         "semantic-release": "semantic-release",
         "start:dbs": "docker-compose -f scripts/mongo-s3-docker-compose.yml up -d",
         "start": "npm run start --prefix packages/h5p-examples",
+        "start:rest": "npm run start --prefix packages/h5p-rest-example-server",
```
It only starts the server, the name should make it clear.

JPSchellenberg committed:
test(contentUserDataGET): add example for GET issue

https://github.com/Lumieducation/H5P-Nodejs-library/issues/1014#issuecomment-968314801

JPSchellenberg committed:
test(contentUserData): revert example

JPSchellenberg committed:
chore(scripts): add h5p-server build script

JPSchellenberg committed:
test(launch): add DEBUG to vscode launch

JPSchellenberg committed:
refactor(script): rename start:rest to start:rest:server

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#discussion_r748820671

JPSchellenberg committed:
refactor(saveFreq): rename saveFreq to contentUserStateSaveInterval

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#discussion_r748820014

JPSchellenberg committed:
feat(contentUserData): delete contentUserData when content is deleted

JPSchellenberg committed:
test(contentUserData): use mock for H5PPlayer.render test

JPSchellenberg committed:
refactor(h5p-examples): add mock implementation to pass build

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
feat(contentUserData): build contentUserDataIntegration in Manager

JPSchellenberg committed:
feat(contentUserData): update interfaces

JPSchellenberg committed:
test(contentUserDataManager): add tests

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
refactor(h5p-examples): remove test/example conentUserDataStorage

JPSchellenberg committed:
refactor(saveContentUserData): add invalidate and preload parameters

JPSchellenberg committed:
feat(h5p-express): add ContentUserDataController

sr258 commented on packages/h5p-express/src/ContentUserDataRouter/ContentUserDataController.ts:
```diff
@@ -0,0 +1,59 @@
+            success: boolean;
+        }>
+    ): Promise<void> => {
+        const { contentId, dataType, subContentId } = req.params;
```
potentially unsafe parameters must be validated / sanitized

sr258 commented on packages/h5p-express/src/ContentUserDataRouter/ContentUserDataController.ts:
```diff
@@ -0,0 +1,59 @@
+        req: IPostContentUserDataRequest,
+        res: express.Response
+    ): Promise<void> => {
+        const { contentId, dataType, subContentId } = req.params;
```
potentially unsafe parameters must be validated / sanitized

JPSchellenberg replied:
I will sanitize these things in the ContentUserDataManager as there will be no need for the user to do so if the user decides to handle the implementation on their own.

sr258 commented on packages/h5p-server/src/ContentManager.ts:
```diff
@@ -21,10 +22,15 @@ const log = new Logger('ContentManager');
+    ) {
         log.info('initialize');
         this.contentStorage = contentStorage;
+        this.contentUserDataStorage = contentUserDataStorage;
```
this.contentUserDataStorage is already set as it is set by the public constructor argument

sr258 commented on packages/h5p-server/src/ContentManager.ts:
```diff
@@ -90,14 +96,27 @@ export default class ContentManager {
         contentId: ContentId,
         user: IUser
     ): Promise<void> {
+        if (this.contentUserDataStorage) {
```
It might be sensible to acquire a lock that is also acquired when the user posts the content state. Otherwise a race condition is possible in which the content state is deleted, created again and then the content is deleted. The effect would be that there are orphaned content states. Alternatively it you could delete the content state after deleting the content.

JPSchellenberg replied:
Ok, I will delete it after the content is deleted.

sr258 commented on packages/h5p-server/src/ContentManager.ts:
```diff
@@ -90,14 +96,27 @@ export default class ContentManager {
+            try {
+                await this.contentUserDataStorage.deleteContentUserData(
+                    contentId,
+                    user
```
The content user data must be deleted for all users, not just the current.

JPSchellenberg replied:
contentUserData is deleted based on the contentId - the user is only passed to check if he has permissions to delete the states.

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * @param contentUserDataStorage The storage object
+     */
+    constructor(contentUserDataStorage: IContentUserDataStorage) {
```
Using a private constructor argument eliminated the need to assign the property and even declare it.

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,154 @@
+            return undefined;
+        }
+
+        if (!this.contentUserDataStorage.loadContentUserData) {
```
Is this really necessary? An implementation wouldn't make much sense without this method, as it's not optional. We don't do this is the library anywhere else.

JPSchellenberg replied:
ok, I will take it out.

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,154 @@
+     * @param dataType Used by the h5p.js client
+     * @param subContentId The id provided by the h5p.js client call
+     * @param user The user who owns this object
+     * @returns the saved state as string
```
or undefined when not found

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,154 @@
+    public async saveContentUserData(
+        contentId: string,
+        dataType: string,
+        subcontentId: string,
```
mismatch with JSDocs name

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,154 @@
+        dataType: string,
+        subcontentId: string,
+        userState: string,
+        invalidate: 0 | 1,
```
boolean would be better here; the library doesn't use 0 or 1 for booleans anywhere else, except if it passes values to the client

JPSchellenberg replied:
I know, but it is what the core sends. But I guess changing it to boolean would not be a problem.

sr258 replied:
The paramters of the core are already changed in other places. We simply have to map the values to the internal types and back. The core is very inconsistent in how booleans are used (0, "0" or false are used) and it's better to have consistence within the library.

sr258 commented on packages/h5p-server/src/index.ts:
```diff
@@ -85,6 +88,7 @@ export {
     streamToString,
     ContentFileScanner,
     ContentTypeCache,
+    ContentUserDataManager,
```
Is it necessary to export it? The other managers are exported as they are used by the html exporter package, but I'm not sure if anyone might want to use the ContentUserDataManager outside of the library. If we export too many classes, it becomes more difficult to change signatures.

JPSchellenberg replied:
If you want to allow users to integrate it under custom routes then yes. It might be necessary to put custom logic into a route and therefore it would be good if you can import the manager and handle the state saving on your own.

sr258 commented on packages/h5p-server/test/ContentManager.test.ts:
```diff
@@ -129,6 +131,33 @@ describe('ContentManager', () => {
+                await expect(
+                    contentManager.deleteContent(contentId, user)
+                ).resolves.toBeUndefined();
+                await expect(
```
It would be better if the test created content states for several users and checks if the content is really deleted for all users.

JPSchellenberg replied:
The contentUserData is deleted based on the contentId. I will add a separate function for a specific user.

sr258 commented on packages/h5p-server/test/ContentUserDataManager.test.ts:
```diff
@@ -0,0 +1,211 @@
+
+import User from './User';
+
+describe('ContentUserDataManager', () => {
```
The tests here mostly check if function of the storage class are called. I think it would be more useful if tests checked actual functionality, like saving a state and retrieving it later. Just checking if calls are passed to the storage class might lead to a false sense of secruity as it produces good line coverage, but doesn't check any real functionality or logic.

JPSchellenberg replied:
I agree with that, but we should think about separating two tests:
unit-test which solely tests the "unit"  and integration tests. If you add an actual storage class you also test the functionality of the storage class. If you want a pure unit test you have to mock everything else.
I will add an integration test as a separate test.

sr258 commented on packages/h5p-server/src/H5PEditor.ts:
```diff
@@ -1112,7 +1123,7 @@ export default class H5PEditor {
             libraryConfig: this.config.libraryConfig,
             postUserStatistics: false,
-            saveFreq: false,
+            saveFreq: this.config.contentUserStateSaveInterval,
```
I guess the editor also needs the url where it can save states

JPSchellenberg replied:
Yep. @otacke mentioned that the editor uses it to store data about tours.

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
fix(contentManager): delete contentUserDataStorage after content

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#discussion_r753821541

JPSchellenberg committed:
refactor(ContentUserDataManager): remove unnecessary code

JPSchellenberg committed:
refactor(ContentUserDataManager): use boolean instead of number

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#discussion_r753822128

JPSchellenberg committed:
refactor(ContentManager): remove unnecessary code

JPSchellenberg committed:
refactor(ContentUserDataManager): use boolean instead of number

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#discussion_r753822128

JPSchellenberg committed:
fix(ContentUserDataManager): sanitize userState before saving

JPSchellenberg committed:
feat(h5pexpress): add ContentUserDataRouter

JPSchellenberg committed:
refactor(h5p-server): add deleteAllContentUserDataforContentId method

JPSchellenberg committed:
refactor(h5p-examples): add reference implementation to h5p-example

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
docs(ContentUserDataStorage): add contentUserDataStorage docs

JPSchellenberg committed:
feat(contentUserData): add setFinished

sr258 commented on docs/usage/h5p-editor-constructor.md:
```diff
@@ -114,3 +114,13 @@ for an implementation sample using the urlGenerator.
+## contentUserDataStorage (optional)
+
+The `contentUserDataStorage` handles saving and loading user states, so users can continue where they left off when they reload the page or come back later. It must implement the `IContentUserDataStorage` interface. You can find a reference implementation in [`/packages/h5p-examples/src/contentUserDataStorage.ts`](/packages/h5p-examples/src/contentUserDataStorage.ts):
+nder `./h5p/content` you can use the sample
```
something is cut off here

sr258 commented on packages/h5p-examples/src/contentUserDataStorage.ts:
```diff
@@ -0,0 +1,135 @@
+        userData = userData.filter(
+            (data) => data.contentId !== contentId && data.userId !== userId
+        );
+        return;
```
redundant; no Promise returned

sr258 commented on packages/h5p-examples/src/contentUserDataStorage.ts:
```diff
@@ -0,0 +1,135 @@
+        requestingUser: IUser
+    ): Promise<void> {
+        userData = userData.filter((data) => data.contentId !== contentId);
+        return;
```
redundant; no Promise returned

sr258 commented on packages/h5p-examples/src/contentUserDataStorage.ts:
```diff
@@ -0,0 +1,135 @@
+            userId: user.id
+        });
+
+        return;
```
redundant

sr258 commented on packages/h5p-examples/src/contentUserDataStorage.ts:
```diff
@@ -0,0 +1,135 @@
+        );
+    }
+
+    public saveContentUserData(
```
Does this method return a Promise if it's not async?

sr258 commented on packages/h5p-examples/src/contentUserDataStorage.ts:
```diff
@@ -0,0 +1,135 @@
+    }
+}
+
+export default new ContentUserDataStorage();
```
The library doesn't use singletons and they are typically considered an anti-pattern. The code that uses the class should construct its own instance with new.

sr258 commented on packages/h5p-examples/src/createH5PEditor.ts:
```diff
@@ -8,6 +8,7 @@ import * as dbImplementations from '@lumieducation/h5p-mongos3';
 import RedisLockProvider from '@lumieducation/h5p-redis-lock';
 import { ILockProvider } from '@lumieducation/h5p-server';
 
+import contentUserDaraStorage from './contentUserDataStorage';
```
Should only import the class and construct it's own object

sr258 commented on packages/h5p-examples/src/express.ts:
```diff
@@ -103,7 +106,9 @@ const start = async (): Promise<void> => {
-        (key, language) => translationFunction(key, { lng: language })
+        (key, language) => translationFunction(key, { lng: language }),
+        undefined,
+        contentUserDataStorage
```
Should get the storage from the editor like libraryStorage

sr258 commented on packages/h5p-express/src/ContentUserDataRouter/ContentUserDataController.ts:
```diff
@@ -0,0 +1,89 @@
+import * as express from 'express';
+
```
I've separated import groups by external imports and internal ones.

sr258 commented on packages/h5p-express/src/ContentUserDataRouter/ContentUserDataController.ts:
```diff
@@ -0,0 +1,89 @@
+    };
+
+    /**
+     * Saving a userState for given contentId, datType and subContentId
```
dat_a_type

sr258 commented on packages/h5p-express/src/ContentUserDataRouter/ContentUserDataExpressRouter.ts:
```diff
@@ -0,0 +1,62 @@
+import { Router } from 'express';
+
```
import groups

sr258 commented on packages/h5p-express/src/ContentUserDataRouter/ContentUserDataExpressRouter.ts:
```diff
@@ -0,0 +1,62 @@
+import ContentUserDataController from './ContentUserDataController';
+/**
+ * This router implements necessary routes for the contentUserData (userState) to work.
+ * If you only want certain routes, you can specify this in the options parameter.
```
JSDocs doesn't match signature

JPSchellenberg replied:
How do I check this again?

sr258 replied:
It's only a divergence between the /** **/ part and the signature.

sr258 commented on packages/h5p-express/src/ContentUserDataRouter/ContentUserDataExpressRouter.ts:
```diff
@@ -0,0 +1,62 @@
+        contentUserDataManager
+    );
+
+    router.get(
```
The router doesn't use the config; I'm not sure if this makes sense, as it's hardcoded by the H5P client anyway; maybe the config shouldn't include the placeholders then

JPSchellenberg replied:
You are right.. I will remove the placeholders

sr258 commented on packages/h5p-express/src/ContentUserDataRouter/ContentUserDataExpressRouter.ts:
```diff
@@ -0,0 +1,62 @@
+        )
+    );
+
+    router.post(
```
Should be configured through setFinishedUrl

sr258 commented on packages/h5p-express/test/ContentUserDataRouter.test.ts:
```diff
@@ -0,0 +1,154 @@
+            user
+        );
+        expect(res.status).toBe(200);
+        // expect(res.body).toBe({ data: mockReturnData, success: true });
```
obsolete?

sr258 commented on packages/h5p-rest-example-server/config.json:
```diff
@@ -6,6 +6,7 @@
     "sendUsageStatistics": false,
     "hubRegistrationEndpoint": "https://api.h5p.org/v1/sites",
     "hubContentTypesEndpoint": "https://api.h5p.org/v1/content-types/",
+    "contentUserDataUrl": "/contentUserData/:contentId/:dataType/:subContentId",
```
It would be good if the rest example also included the user state feature, as I think it is used a lot.

sr258 commented on packages/h5p-server/src/implementation/H5PConfig.ts:
```diff
@@ -98,8 +99,9 @@ export default class H5PConfig implements IH5PConfig {
-    public setFinishedUrl: string = '/setFinished';
+    public setFinishedUrl: string = '/contentUserData/setFinished';
     public siteType: 'local' | 'network' | 'internet' = 'local';
+    public contentUserStateSaveInterval: number = 30; // the interval to save the contentUserData in seconds
```
I think it would make sense to change this into milliseconds and convert the interval into seconds for H5P. That way the configuration isn't inconsistent.

sr258 commented on packages/h5p-server/src/ContentManager.ts:
```diff
@@ -90,15 +94,31 @@ export default class ContentManager {
+            }
+        }
+
+        return;
```
redundant

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+import {
```
internal import group comes after external imports

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+        requestingUser: IUser
+    ): Promise<void> {
+        if (this.contentUserDataStorage) {
+            log.info(
```
Should be debug

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+                requestingUser
+            );
+        }
+        return;
```
redundant

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+        requestingUser: IUser
+    ): Promise<void> {
+        if (this.contentUserDataStorage) {
+            log.info(`deleting all contentUserData for ContentId ${contentId}`);
```
Should be debug

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+                requestingUser
+            );
+        }
+        return;
```
redundant

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+            return undefined;
+        }
+
+        log.info(
```
should be debug (and also log calls below)

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+        );
+
+        const sortedStates = states.sort(
+            (a, b) => Number(a.subContentId) - Number(b.subContentId)
```
Is this a save conversion? Is the subcontent id always a number?

JPSchellenberg replied:
As far as I reconstructed it the subContentId equals the index in the array that is in [contentUserData](https://github.com/Lumieducation/H5P-Nodejs-library/blob/feat/1014-contentUserData/packages/h5p-server/src/H5PPlayer.ts#L391) field in the Integration object. Therefore it should always be a number.

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    /**
+     * Saves data when a user completes content. T
```
T seems like a leftover

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+     * Saves data when a user completes content. T
+     * @param contentId The content id to delete.
+     * @param score the score the user reached as an integer
+     * @param maxScore the maximal score of the content
```
maximum

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+
+        if (!sanitizedUserState || sanitizedUserState === '') {
+            log.error(`no userState provided for ${contentId}`);
+            throw new Error('no-userState');
```
Is this a message that should be sent to the user? There's the H5PError class for that. The convention for errors used in the rest of the library is 'no-user-state'

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+ * The ContentUserDataManager takes care of saving user data and states. It only contains storage-agnostic functionality and
+ * depends on a ContentUserDataStorage object to do the actual persistence.
+ */
+export default class ContentUserDataManager {
```
The finished state should also be retreived somehow: all states for the contentId and the individual state for a single user

JPSchellenberg replied:
But wouldn't that be some kind of round-trip? The user would have to implement a corresponding method in the ContentUserDataStorage. I can't think of a scenario where a user would use a method on the manager class to retrieve data that way or am I missing something?

sr258 replied:
Yes, it can be left out in the manager, but should be included in the interface.

sr258 commented on packages/h5p-server/src/H5PPlayer.ts:
```diff
@@ -60,7 +62,8 @@ export default class H5PPlayer {
         }).t,
-        private options?: IH5PPlayerOptions
+        private options?: IH5PPlayerOptions,
+        contentUserDataStorage?: IContentUserDataStorage
```
The options should come last; I would opt for a breaking change and collect all storage classes in a separate object

JPSchellenberg replied:
I will make a separate PR for that when this PR is merged.

sr258 commented on packages/h5p-server/src/types.ts:
```diff
@@ -543,6 +544,112 @@ export interface IUser {
+ * Implementations can implement the IContentUserDataStorage interface and pass it to the constructor of
+ * H5PEditor or H5PPlayer if they wish to keep the user state and allows users to continue where they left off.
+ */
+export interface IContentUserDataStorage {
```
The interface should also include capbilities to delete the user data for a specific user only across all content objects and to list all user data for a specific user. (GDPR compliance)

JPSchellenberg replied:
But from where in the library should it be called?

sr258 replied:
The implementation can do it on its own. It constructs the storage, so it can also call the methods.

sr258 commented on packages/h5p-server/src/ContentUserDataManager.ts:
```diff
@@ -0,0 +1,211 @@
+            `saving contentUserData for user with id ${user.id} and contentId ${contentId}`
+        );
+
+        const sanitizedUserState = sanitizeHtml(userState);
```
Does this work for JSON object? It's JSON string after all...

JPSchellenberg replied:
For what would you sanitise the object? The sanitizeHtml removes all javascript code, but is there something I can't think of?

sr258 replied:
I wouldn‘t sanitize it at all. It‘s inherently unsafe, but we can’t do anything about it. The other args might need validation, but I wouldn‘t do it for the arbitrary data in the actual state.

JPSchellenberg replied:
Alright. How would validate the other arguments? Just by checking the typeof?

sr258 replied:
Yes, that‘ll do. I‘d check the number and boolean types in particular. These tend to be inconsistent in H5P.

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
fix(ContentUserDataManager): throw H5pError instead of regular error

JPSchellenberg committed:
fix(contentUserStateSaveInterval): change from seconds to milliseconds

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#discussion_r760102698

JPSchellenberg committed:
refactor(code): remove redundant return statements

JPSchellenberg committed:
refactor(code): reorder import statements

JPSchellenberg committed:
refactor(log): change to debug from info

JPSchellenberg committed:
refactor(code): remove typos, reorder imports, remove redundant code

JPSchellenberg committed:
refactor(h5p-examples): don't use singleton

JPSchellenberg committed:
fix(H5PPlayer): use contentUserStateSaveInterval in milliseconds

JPSchellenberg committed:
refactor(code): fix typos

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
fix(contentUserDataManager): saveContentUserData: check arguments

otacke committed:
fix(h5p-examples): correct json number declaration (#1991)

JPSchellenberg committed:
Merge remote-tracking branch 'origin/feat/1014-contentUserData' into feat/1014-contentUserData

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
feat(listContentUserDataByUserId): add listContentUserDataByUserId

JPSchellenberg committed:
style(jsdocs): add divergence

sr258 committed:
refactor: formatting and minor issues

sr258 committed:
refactor: fixed imports

sr258 committed:
test: corrected test

sr258 committed:
refactor: decreased content user state save interval

sr258 committed:
Merge remote-tracking branch 'origin/master' into feat/1014-contentUserData

sr258 committed:
test: fix test

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
fix(contentUserDataManager): remove userData when invalidate is true

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#issuecomment-1050627214

JPSchellenberg committed:
fix(contentUserData): generate integration only when preload is true

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#issuecomment-1050627214

JPSchellenberg committed:
feat(FileContentUserDataStorage): add FileContentUserDataStorage

JPSchellenberg committed:
feat(contentUserData): include contentUserData in rest-example

JPSchellenberg committed:
feat(contentUserData): include in rest example

JPSchellenberg committed:
test(contentUserDataRouter): fix test

JPSchellenberg committed:
chore(contentUserDataStorage): add json to gitignore

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
build(h5p-shared-state-server): build h5p-shared-state-server on install

JPSchellenberg committed:
Merge branch 'master' into feat/1014-contentUserData

JPSchellenberg committed:
fix(contentUserData): delete invalid contentUserData if content changes

https://github.com/Lumieducation/H5P-Nodejs-library/pull/1886#issuecomment-1059275916

JPSchellenberg committed:
refactor(contentUserData): rename saveContentUserData

saveContentUserData-method is renamed to createOrUpdateContentUserData

JPSchellenberg committed:
feat(mongos3): add MongoContentUserDataStorage

sr258 committed:
Merge remote-tracking branch 'origin/master' into feat/1014-contentUserData

sr258 committed:
refactor: made namings more consistent

sr258 committed:
refactor: minor cleanup

sr258 committed:
fix: examples work

sr258 committed:
test: use snapshot instead of inline HTML

sr258 committed:
feat: corrected and extended MongoContentUserDataStorage

sr258 committed:
fix: corrected FileContentUserDataStorage signature

sr258 committed:
test: added more tests

sr258 committed:
test: more tests

sr258 committed:
feat: delete finished data when content is deleted

sr258 committed:
feat: reimplemented FileContentUserDataStorage to work with directories

sr258 committed:
test: added tests for FileContentUserDataStorage

sr258 committed:
test: renamed generalized test file

sr258 committed:
fix: missing user data doesn't return 404

sr258 committed:
fix: corrected CSFR token generation + own route for setFinished

sr258 committed:
fix: protected FileContentUserDataStorage against attacks

sr258 committed:
test: corrected tests

sr258 committed:
fix: corrected config to load falsy settings

sr258 committed:
fix: improved filename validation

sr258 committed:
feat: routes are closed when feature disabled in config

sr258 committed:
feat: added CSRF tokens to rest example & fixed bugs

sr258 committed:
fix: fileContentUserDataStorage saves more than one entry

sr258 committed:
test: corrected test

sr258 committed:
feat: improved  CSRF

sr258 committed:
Merge remote-tracking branch 'origin/master' into feat/1014-contentUserData

sr258 committed:
refactor: correct shared state example

sr258 committed:
refactor: cleanup

sr258 committed:
docs: added docs

sr258 committed:
Merge remote-tracking branch 'origin/master' into feat/1014-contentUserData
